blog/src/Auth/Form/ChangePasswordForm.php

```diff
@@ -0,0 +1,150 @@
+use Yiisoft\Validator\Rule\Callback;
+use Yiisoft\Validator\Rule\Length;
+use Yiisoft\Validator\Rule\Required;
+use Yiisoft\Validator\ValidatorInterface;
+use Yiisoft\Validator\RulesProviderInterface;
+
+final class ChangePasswordForm extends FormModel implements RulesProviderInterface
+{
+    private string $login = '';
+    private string $password = '';
+    private string $newPassword = '';
+    private string $newPasswordVerify = '';
+    
+    public function __construct(
+        private AuthService $authService,
+        private ValidatorInterface $validator,
+        private TranslatorInterface $translator,
+        private UserRepository $userRepository,
+    ) {
+    }
+    
+    public function change(): bool
+    {
+        if ($this->validator->validate($this)->isValid()) {
+            $user = $this->userRepository->findByLogin($this->getLogin());
+            if (null!==$user) {
+              $user->setPassword($this->getNewPassword());
+              // The cookie identity auth_key is regenerated on logout
+              // Refer to ResetController reset function
+              $this->userRepository->save($user);
+              return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @return string[]
+     *
+     * @psalm-return array{login: string, password: string, newPassword: string, newPasswordVerify: string}
+     */
+    public function getAttributeLabels(): array
+    {
+        return [
+            'login' => $this->translator->translate('layout.login'),
+            'password' => $this->translator->translate('layout.password'),
+            'newPassword' => $this->translator->translate('layout.password.new'),
+            'newPasswordVerify' => $this->translator->translate('layout.password-verify.new'),
+        ];
+    }
+
+    /**
+     * @return string
+     *
+     * @psalm-return 'Change'
+     */
+    public function getFormName(): string
+    {
+        return 'ChangePassword';
+    }
+    
+    public function getLogin(): string
+    {
+        return $this->login;
+    }
+    
+    public function getPassword(): string
+    {
+        return $this->password;
+    }
+        
+    public function getNewPassword(): string
+    {  
+        return $this->newPassword;
+    }
+    
+    public function getNewPasswordVerify() : string
+    {
+        return $this->newPasswordVerify;
+    }
+    
+    public function getRules(): array
+    {
+        return [
+            'login' => [
+                new Required(),
+                new Length(min: 1, max: 48, skipOnError: true),
+                function (mixed $value): Result {
+                    $result = new Result();
+                    if ($this->userRepository->findByLogin((string)$value) == null) {
+                        $result->addError($this->translator->translate('validator.user.exist.not'));
+                    }
+                    return $result;
+                },
+            ],
```

Comment:
Ideally, that should be implemented by catching a database exception on inserting a record because `SELECT` and then `INSERT` aren't atomic and, thus, can cause an exception anyway if a value is inserted by another user in between.